### PR TITLE
Replace jshint and jscs with eslint

### PR DESCRIPTION
**Pull request by gilligan**
### what
As mentioned on gitter I would like to replace `jshint` and `jscs` with `eslint`. 

### why
eslint is flexible and powerful enough to replace the functionality of both tools. Even if rules beyond current capabilities might be required at some point there is a nice plugin architecture which would allow for custom rules. Apart from that jshint is somewhat archaic.

### status
I created a `.eslintrc` based on the existing linting configurations. The configuration was extended with rules such as http://eslint.org/docs/rules/wrap-iife and http://eslint.org/docs/rules/one-var plus a various rules for catching obvious mistakes such as dangling commas or duplicate function arguments. Obviously all sources were fixed to adhere to the rules and `npm run lint` reports no errors.

### CODE REVIEW

**Comment by buzzdecafe** [comment]
fine with me, i have no great love for jscs or jshint. so if we can use 1 tool, and reduce our conf-file count, then let's.

File: src/internal/_flatCat.js [hunk 1]
```diff
@@ -29,4 +29,4 @@ module.exports = (function() {
       }
     };
   };
-}());
+})();
```
Comment (CrossEye) [inline src/internal/_flatCat.js]: I'm curious as to the rationale for this.  Is this simply the eslint house style?  Is it configurable?

I admit to a minor preference for the other style, `(function() {/* ... */}())` over `(function() {/* ... */})()`, but it's a fairly minor things.  Still, I don't want to change this for no good reason.  And I think "that's what the linter prefers" is no good reason.  If someone can make a compelling argument, or if eslint doesn't even allow my preferred syntax, then I could see just going with the flow.

Comment (CrossEye) [inline src/internal/_flatCat.js]: Never mind.  Looked it up and see this non-default preference in the config file above.  Will add a comment there.

File: .eslintrc [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      2,
+      "last"
+    ],
+    "curly": [
+      2,
+      "all"
+    ],
+    "dot-notation": 0,
+    "eol-last": 2,
+    "wrap-iife": [2, "inside"],
```
Comment (CrossEye) [inline .eslintrc]: That's causing a number of tweaks that seem to have little reason.  Were there more if it was the other way?

I'd rather do "outside", although it's a minor preference.

**Comment by CrossEye** [comment]
I'd be quite happy to have this.  I do have a caveat about one of the flags, but I think this is worth doing, overall.

File: src/uniqBy.js [hunk 3]
```diff
@@ -36,12 +36,12 @@ module.exports = _curry2(/* globals Set */ typeof Set === 'undefined' ?
...
-        result        = [],
-        nullExists    = false,
-        negZeroExists = false,
-        idx           = 0;
+      applied       = [], appliedItem, item,
+      prevSetSize   = 0,  newSetSize,
+      result        = [],
+      nullExists    = false,
+      negZeroExists = false,
+      idx           = 0;
```
Comment (scott-christopher) [inline src/uniqBy.js]: The `=` here either needs realigning or the extra whitespace used for alignment should be removed.

**Comment by scott-christopher** [comment]
It might be worth standardising on the use of one var per block vs one var per variable while we're at it too. It looks like the `one-var` eslint config covers this, which I'd propose using:
```
"one-var": [2, {
    "uninitialized": "always", // Exactly one declaration for uninitialized variables per function
    "initialized": "never" // Exactly one declarator per initialized variable declaration per function
}]
```

File: .eslintrc (hunk 2 above)
Comment (gilligan) [inline .eslintrc]: Neither inside nor outside was used consistently so I basically just picked one. I'm totally fine with changing it to "outside" if that is preferred.

File: src/uniqBy.js (hunk 3 above)
Comment (gilligan) [inline src/uniqBy.js]: Everything is indented by 2 spaces.

Comment (scott-christopher) [inline src/uniqBy.js]: Sorry, I meant the `=` before `new Set()` is no longer vertically aligned with the other `=`.

**Comment by gilligan** [comment]
I now switched the `iife` config so the expected syntax is now : 
```js
(function () { ... }()); 
```
As you can see there were number of files where it was used the other way round.

**Comment by gilligan** [comment]
@scott-christopher I added the `one-var` rule as suggested by you.

File: .eslintrc [hunk 4]
```diff
@@ -0,0 +1,87 @@
+{
+  "env": {
+      "node": true,
+      "mocha": true
```
Comment (davidchambers) [inline .eslintrc]: :triangular_ruler: We should lint __.eslintrc__. ;)

File: src/empty.js [hunk 5]
```diff
@@ -40,7 +40,7 @@ module.exports = _curry1(function empty(x) {
     _isObject(x) ?
       {} :
     _isArguments(x) ?
-      (function() { return arguments; }()) :
+      (function() { return arguments; })() :
```
Comment (davidchambers) [inline src/empty.js]: This change should be reverted, shouldn't it?

**Comment by gilligan** [comment]
While adjusting the sources for `one-var` I realised that sometimes there is a blank line after the var statements block and sometimes there is not. Is this something you would like to see standardised the one or the other way @davidchambers @scott-christopher ?

**Comment by davidchambers** [comment]
This needn't be standardized.

**Comment by gilligan** [comment]
I think the following rules could be enabled because they are pretty much uncontroversial and just find silly mistakes : 

http://eslint.org/docs/rules/comma-dangle
http://eslint.org/docs/rules/no-debugger 
http://eslint.org/docs/rules/no-dupe-args 
http://eslint.org/docs/rules/no-dupe-keys 
http://eslint.org/docs/rules/no-duplicate-case 
http://eslint.org/docs/rules/no-extra-semi
http://eslint.org/docs/rules/no-unreachable

There is no good reason to allow any of what the above checks I suppose ?

**Comment by davidchambers** [comment]
Go ahead and enable those rules, @gilligan. :)

**Comment by gilligan** [comment]
Added the aforementioned rules. There were some dangling commas at least ;--} Are there any other modifications anyone would like to see to the rules ? Otherwise I'd squash commits and make this a proper PR.

**Comment by gilligan** [comment]
stashed the commits, updated title and description. The changes are of course pretty much all over the place but all of them very minimal.

File: .eslintrc [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    "quotes": [ 2, "single", "avoid-escape" ],
+    "comma-dangle": [ 2, "never" ],
+    "no-debugger": 2,
+    "no-dupe-args": 2,
+    "no-dupe-keys": 2,
+    "no-duplicate-case": 2,
+    "no-extra-semi": 2,
+    "no-unreachable": 2,
+  }
+}
```
Comment (davidchambers) [inline .eslintrc]: Could we make the formatting of this file consistent? We use `[a, b, c]` in some places and `[ a, b, c ]` in others. You could use `JSON.stringify(x, null, 2)` to normalize the formatting.

**Comment by davidchambers** [comment]
:deciduous_tree:

**Comment by CrossEye** [comment]
:fireworks:

**Comment by davidchambers** [comment]
@gilligan, can we remove __.jscsrc__, __.jshintignore__, and __.jshintrc__?

**Comment by CrossEye** [comment]
I believe we can.

**Comment by gilligan** [comment]
Oh, yes we can/should indeed. Forgot that :/

**Comment by davidchambers** [comment]
Mind opening a pull request, @gilligan?

**Comment by gilligan** [comment]
@davidchambers #1559 

I just now realised that there were also still plenty of jshint switches across various test files ;(
The PR removes all of them.

**Comment by davidchambers** [comment]
Thanks!

**Pull request merged**